### PR TITLE
LUCENE-8933: Validate JapaneseTokenizer user dictionary entry

### PR DESCRIPTION
**Pull request by mocobeta**
# Description

This adds a format check to Kuromoji user dictionary if the concatenated segment is same as its surface form. See https://issues.apache.org/jira/browse/LUCENE-8933

### CODE REVIEW

File: lucene/analysis/kuromoji/src/java/org/apache/lucene/analysis/ja/dict/UserDictionary.java [hunk 1]
```diff
@@ -104,6 +104,8 @@ public int compare(String[] left, String[] right) {
     long ord = 0;
     
     for (String[] values : featureEntries) {
+      String surface = values[0].replaceAll(" ", "");
```
Comment (msokolov) [inline lucene/analysis/kuromoji/src/java/org/apache/lucene/analysis/ja/dict/UserDictionary.java]: maybe replace all white space here (ie including tabs) by using \s?

Comment (msokolov) [reply lucene/analysis/kuromoji/src/java/org/apache/lucene/analysis/ja/dict/UserDictionary.java]: ah never mind my question, I see from the email discussion that is the case: we shouldn't allow this kind of entry

Comment (mocobeta) [reply lucene/analysis/kuromoji/src/java/org/apache/lucene/analysis/ja/dict/UserDictionary.java]: > maybe replace all white space here (ie including tabs) by using \s?

Thanks, fixed.

File: lucene/analysis/kuromoji/src/test/org/apache/lucene/analysis/ja/dict/UserDictionaryTest.java [hunk 2]
```diff
@@ -77,4 +78,19 @@ public void testRead() throws IOException {
     UserDictionary dictionary = TestJapaneseTokenizer.readDict();
     assertNotNull(dictionary);
   }
+
+  @Test(expected = RuntimeException.class)
```
Comment (jimczi) [inline lucene/analysis/kuromoji/src/test/org/apache/lucene/analysis/ja/dict/UserDictionaryTest.java]: Can you use `expectThrows` and check that the message is correct ?

File: lucene/analysis/kuromoji/src/test/org/apache/lucene/analysis/ja/dict/UserDictionaryTest.java [hunk 3]
```diff
@@ -77,4 +78,19 @@ public void testRead() throws IOException {
...
+  public void testReadInvalid1() throws IOException {
+    // the concatenated segment must be the same as the surface form
+    String invalidEntry = "日経新聞,日本 経済 新聞,ニホン ケイザイ シンブン,カスタム名詞";
+    UserDictionary dictionary = UserDictionary.open(new StringReader(invalidEntry));
+  }
+
+  @Test(expected = RuntimeException.class)
```
Comment (jimczi) [inline lucene/analysis/kuromoji/src/test/org/apache/lucene/analysis/ja/dict/UserDictionaryTest.java]: same here

File: lucene/analysis/kuromoji/src/test/org/apache/lucene/analysis/ja/dict/UserDictionaryTest.java (hunk 2 above)
Comment (mocobeta) [reply lucene/analysis/kuromoji/src/test/org/apache/lucene/analysis/ja/dict/UserDictionaryTest.java]: I think you mean `assertThrows` that was introduced from JUnit 4.13
https://junit.org/junit4/javadoc/latest/org/junit/Assert.html#assertThrows(java.lang.String,%20java.lang.Class,%20org.junit.function.ThrowingRunnable)

Lucene uses junit 4.12, so we cannot use it (Am I missing something?).

**Comment by mocobeta** [comment]
Thanks, I will merge it soon.

File: lucene/analysis/kuromoji/src/test/org/apache/lucene/analysis/ja/dict/UserDictionaryTest.java (hunk 2 above)
Comment (mocobeta) [reply lucene/analysis/kuromoji/src/test/org/apache/lucene/analysis/ja/dict/UserDictionaryTest.java]: Sorry, I missed LuceneTestCase#expectThrows. Will fix this in another patch.

Comment (mocobeta) [reply lucene/analysis/kuromoji/src/test/org/apache/lucene/analysis/ja/dict/UserDictionaryTest.java]: Fixed in https://github.com/apache/lucene-solr/pull/830

**Pull request merged**